### ISSUE TITLE
extras_require options do not work with python3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7807]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)| @derwolfe reported|
|-|-|
|Trac ID|trac#7807|
|Type|defect|
|Created|2015-03-11 00:07:47Z|
|Branch|https://github.com/twisted/twisted/tree/7807-rodrigc-extrasrequire-py3|

The extras_require options introduced with ticket [#3696](https://github.com/twisted/twisted/issues/3696) fail when python3 is used. This can be reproduced by taking the following steps:

The following steps assuming python 3.3 is installed

```
# cd into root directory of twisted repostory
$ cd ./twisted

# create a python3 virtualenv
$ pyenv-3.3 venv

# activate the environment
$ . ./venv/bin/activate

# install an editable version of twisted from the local source with the `dev` option
$ pip install -e ".[dev]"

Obtaining file:///Users/chris/Code/twisted
    file twisted.py (for module twisted) not found
    file twisted/internet.py (for module twisted.internet) not found
    file twisted/internet/test.py (for module twisted.internet.test) not found
    file twisted/names.py (for module twisted.names) not found
    file twisted/names/test.py (for module twisted.names.test) not found
    file twisted/protocols.py (for module twisted.protocols) not found
    file twisted/protocols/test.py (for module twisted.protocols.test) not found
    file twisted/python.py (for module twisted.python) not found
    file twisted/python/test.py (for module twisted.python.test) not found
    file twisted/test.py (for module twisted.test) not found
    file twisted/trial.py (for module twisted.trial) not found
    file twisted/trial/test.py (for module twisted.trial.test) not found
    file twisted/web.py (for module twisted.web) not found
    file twisted/web/test.py (for module twisted.web.test) not found
  Exception:
  Traceback (most recent call last):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2501, in requires
      deps.extend(dm[safe_extra(ext)])
  KeyError: 'dev'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/basecommand.py", line 232, in main
      status = self.run(options, args)
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/commands/install.py", line 339, in run
      requirement_set.prepare_files(finder)
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/req/req_set.py", line 436, in prepare_files
      req_to_install.extras):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2504, in requires
      "%s has no such extra feature %r" % (self, ext)
  pip._vendor.pkg_resources.UnknownExtra: Unknown 15.0.0 has no such extra feature 'dev'```{
# cd into root directory of twisted repostory
$ cd ./twisted

# create a python3 virtualenv
$ pyenv-3.3 venv

# activate the environment
$ . ./venv/bin/activate

# install an editable version of twisted from the local source with the `dev` option
$ pip install -e ".[dev]"

Obtaining file:///Users/chris/Code/twisted
    file twisted.py (for module twisted) not found
    file twisted/internet.py (for module twisted.internet) not found
    file twisted/internet/test.py (for module twisted.internet.test) not found
    file twisted/names.py (for module twisted.names) not found
    file twisted/names/test.py (for module twisted.names.test) not found
    file twisted/protocols.py (for module twisted.protocols) not found
    file twisted/protocols/test.py (for module twisted.protocols.test) not found
    file twisted/python.py (for module twisted.python) not found
    file twisted/python/test.py (for module twisted.python.test) not found
    file twisted/test.py (for module twisted.test) not found
    file twisted/trial.py (for module twisted.trial) not found
    file twisted/trial/test.py (for module twisted.trial.test) not found
    file twisted/web.py (for module twisted.web) not found
    file twisted/web/test.py (for module twisted.web.test) not found
  Exception:
  Traceback (most recent call last):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2501, in requires
      deps.extend(dm[safe_extra(ext)])
  KeyError: 'dev'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/basecommand.py", line 232, in main
      status = self.run(options, args)
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/commands/install.py", line 339, in run
      requirement_set.prepare_files(finder)
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/req/req_set.py", line 436, in prepare_files
      req_to_install.extras):
    File "/Users/chris/.pyenv/versions/3.3.5/lib/python3.3/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2504, in requires
      "%s has no such extra feature %r" % (self, ext)
  pip._vendor.pkg_resources.UnknownExtra: Unknown 15.0.0 has no such extra feature 'dev'


Attachments:

* [extras-require-py3-7807-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/440/440342f4a9f7067fb1ce502ed7e7d02656ccefab/122819dbbeab6a0caeca6ed7a02db18b25a84839.patch) (624 bytes) - added by herrwolfe on 2015-03-11 00:35:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7807 7807
type__defect defect
reporter__herrwolfe herrwolfe
priority__normal normal
milestone__ 
branch__7807_rodrigc_extrasrequire_py3 7807-rodrigc-extrasrequire-py3
branch_author__hawkowl__herrwolfe hawkowl, herrwolfe
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1426032467382508 1426032467382508
changetime__1467629731230803 1467629731230803
version__None None
owner__rodrigc rodrigc
cc__kmike
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_1"></a>|@derwolfe commented|
|-|-|

Several of the extra options are not yet python3 compatible.

- The `dev` option fails because twistedchecker's setup.py file doesn't appear to be compatible with python3.
- `tls` succeeds
- `conch` succeeds
- `soap` fails due to its dependency on wstools
- `serial` succeeds
- `osx_platform`, I couldn't try. The library states that is compatible with python3. But, as this install pyobjc and the `soap` options (in addition to `tls`, `conch`, and `serial`) I expect it to fail due to the wstools failure.
- `windows_platform` - same issue as with `osx_platform`; the pypiwin32 library supports py3, but the option will fail due to wstools.

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_2"></a>|@derwolfe commented|
|-|-|

The patch https://twistedmatrix.com/trac/attachment/ticket/7807/extras-require-py3-7807-1.patch adds the extras_require options to the setup3.py file. It makes no effort to remove python3 incompatible dependencies.

As of right now, not all of the options will successfully install on python3. As far as I can tell, the failing options are `soap`, `dev`, `osx_platform` and `windows_platform`. 

The failures are
- `dev` fails because of  `twistedchecker` 
- `soap` fails because of `wstools`
- `osx_platform`, `windows_platform` fail because of the `soap` option

If the patch is taken as is, users can install most of the dependencies. But, the install will fail if an option has dependencies that incompatible with python3.

Should a different set of options be available on py3? This seems workable, but it would require extra effort to maintain the list until all of the dependencies are py3 compatible.

Thank you.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @derwolfe|
|-|-|

Many thanks for working on this and sorry for the delay.

```
args['extras_require']=_EXTRAS_REQUIRE
```

should be

```
args['extras_require'] = _EXTRAS_REQUIRE
```

-------

The patch also needs a newsfile

-------

I think that is important to have at least the `.dev` environment installable on py3 as otherwise, what is the point of having this patch and how can we test it in an automated waw ?

I will try to port twistedchecker on python3 as I think that we are close... I have already tried this but got stuck at FilePath, which now is ported.

For wstools there is this ticket https://github.com/pycontribs/wstools/issues/3 

I think that we can skip wstools on python3, so yes, we should have different dependencies (subset) for Python3 as we also have support different features (a subset) for Python3.

----

In the same time I think that Twisted is too big... why not have the SOAP support as an external project ?
But this is another discussion and another ticket.

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_5"></a>|@derwolfe commented|
|-|-|

Thanks for your review adiroiban.

I will work on your comments and add the review flag back once I have made your requested changes.

**Comment by twisted-trac** [comment]
|<img alt="kmike's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|kmike commented|
|-|-|

> I think that is important to have at least the .dev environment installable on py3 as otherwise, what is the point of having this patch and how can we test it in an automated way ?

I was hit by this issue when trying to install twisted[tls] in Python 3.4, so IMHO it is useful even without [dev].

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

(In [46952]) Branching to extras-py3-7807.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc commented|
|-|-|

Patch updated.
See:
https://github.com/twisted/twisted/pull/287

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_10"></a>|@Lukasa set owner to @rodrigc|
|-|-|

One really tiny nit of code review in the diff, but otherwise I'm happy with this. Resubmit for review when that's dealt with, and feel free to assign directly back to me.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_11"></a>|@Lukasa commented|
|-|-|

Given some further discussion on the ticket, I'm happy with this as-is. Feel free to merge.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|GitHub set status to `closed`|
|-|-|

In changeset 40c4268b4c8a2dffc8f8cc2a9ed390d27ba7bcd2
```
#!CommitTicketReference repository="" revision="40c4268b4c8a2dffc8f8cc2a9ed390d27ba7bcd2"
Merge pull request #287 from twisted/7807-rodrigc-extrasrequire-py3

Author: rodrigc
Reviewer: Lukasa
Fixes: #7807

Fix on Python 3:  pip install -e ".[dev]"
```
